okteto/okteto
# Implement new event ImageBuild

# Proposed changes

Resolves https://okteto.atlassian.net/browse/LAKE-46

We want to track every image built using okteto. This means any other build run from the build command itself, or any other command such as deploy, up, destroy, stack that requires the image to be built prior to execute itself.

We want to measure the cachehit when looking into the image and skipping the build in order to make some improvements when using the commit of the build context instead on the commit at the repository itself. For this to happen the metadata and tracking of the event is placed at the V2 Builder Build funcion.

Proposed changes are the following:
- create a new event at `analytics` pkg where we locate the props to send and the event name. Required data is sent hashed and its the tracker who is hashing the information.
- Inject the analyticsTracker into v2 Builder Build though a new property at OktetoBuilder struct `analyticsTracker`. This has caused some refactor around the use of v2 Builder from other commands, as the injection of the tracker is done at `NewBuilderFromScratch`
  - deploy, destroy, init, up, stack commands now include also an analyticsTracker interface and tracker injection so when the build is done we can track this.
- retrieving the metadata for the event has resulted on some refactoring
  - `meta.RepoURL` - repository was not available under the `bc.Confing` object, so made the changes to make it available
  - `meta.RepoHash` - is using the build hash, changed the function that calculates this from being inside the `Config` object into a local function renamed to `getBuildHashFromCommit`
- refactor of OktetoBuilder.areImagesBuilt into local function areAllServicesBuilt using a control map instead of the buildImages property aat the OktetoBuilder struct. The control map is only used locally on the loop so there is no need to tie it to the struct. Also created a local function so its more readeable the condtion `isServiceBuilt` which looks at the control map to check it the service has been built
- refactor around the imageTagger as i found it really difficult to follow what was going on at each function. Mainly the name "sha" used to refer to "tag" on an image referece. 

> The SHA of an image is a unique identifyer and at a docker image is reference like [name]@sha256:[sha] ; on the other hand a tag is a label assigned to the image, which is referenced as [name]:[tag]. Multiple image references with different tags, can be referencing a single SHA.

  - using `tag` when refering to the tag of an image reference: [name]:[tag]
  - rename function `tag` into `getServiceImageReference` as the function returns the image reference in the format [name]:[tag]
  - rename `getImageFromTmpl` into `useReferenceTmplate`
  - rename `getPossibleTags` into `getImageReferencesForTagWithDefaults` as the function returned the references with the tag provided and the default okteto tags
  - rename `getPossibleHashImages` into `getImageReferencesForTag` as the function returned the references wth the tag provided, being this a hash or not.
  - rename `getImageFromTmplWithVolumesAndSHA` into `useReferenceTemplateWithVolumes` as the tag provided could be a hash or a regular tag, either way sha is not the type of value passed here.
  
- refactor at `imageChecker` as in the tagger, i found it difficult so i renamed and added couple comments in order to make it more readeable
  - removed dependency with `GetBuildHash` and now is a param on the functions that need it
  - renaming of sha to tag, and tag to reference
  - rename `getImageSHA` into `lookupReferenceWithDigest` as this is the value returned. An image reference with digest has the following format [name]:[tag]@sha256:[sha/digest]
  - rename `checkIfCommitHashIsBuilt` into `checkIfBuildHashIsBuilt` as the value that is being using for tagging the iamges is the buildHash and not the commit itself, this can be missleading.
  - rename `getImageDigestFromAllPossibleTags` to `getImageDigestReferenceForService` as this is returning the references of images for a service.

- refactor at `GetServicesToBuild`. This function is the link of the build command with other commands. It returns a list of services that need to be built, and for that it checks if they are already built.
  - rename tha chan from `toBuild` into `toBuildCh`
  - rename `checkServicesToBuild` into `checkServiceToBuild` as the function itself checks one by one so its no sense to use the plural
  - inject the `buildHash` into refactored `getImageDigestReferenceForService`

## How to validate

- Using cloud/staging environment, pull first approach is disabled, testing can be done using this context to check that the events are generated and sent
- Using dev cluster, check the metadata for pull first aproach is correctly retrieved

- Create a new namespace and run okteto build for a repository/project to check that on every build an event is sent with the metadata.
- Go to mixpanel to check the event imagebuild filtering by version (local binary commit) or add a printlin at the TrackImageBuild function to see the event being sent with the props if using ur dev cluster.
- Run up, deploy, destroy, stack and all the commands that builder is injected to check that the images are built and the event is sent


## Additional context for reviewers

Attaching diagrams used for the analysis and implementation of the event on how build is done and related with other commands

![Builders](https://github.com/okteto/okteto/assets/40767058/f84fc7c0-da70-4dd3-b3aa-f9e2d2ff3b4a)

![Commands using build](https://github.com/okteto/okteto/assets/40767058/815b7309-c06d-4046-ba22-80820cf18dfc)


## CLI Quality Reminders 🔧

For both authors and reviewers:

- Scrutinize for potential regressions
- Ensure key automated tests are in place
- Build the CLI and test using the validation steps
- Assess Developer Experience impact (log messages, performances, etc)
- If too broad, consider breaking into smaller PRs
- Adhere to our [code style](https://github.com/okteto/okteto/blob/master/docs/code-style.md) and [code review](https://github.com/okteto/okteto/blob/master/docs/code-review.md) guidelines
